File: glisz/misja1/mission1.cpp
#include <iostream>
#include <vector>
#include <random>
#include <deque>
#include <unordered_set>
#include <cmath>
#include <fstream>
#include <functional>
#include <algorithm>
#include<memory>

class Graph {
public:
    std::vector<std::vector<int>> edges;
    int n;

    Graph(int n) : n(n) {
        possible_edges.reserve(n * (n - 1) / 2);
        for (int u = 0; u < n; ++u) {
            for (int v = u + 1; v < n; ++v) {
                possible_edges.emplace_back(std::make_pair(u, v));
            }
        }
        edges.resize(n);
    }

    void add_edge(int u, int v) {
        edges[u].push_back(v);
        edges[v].push_back(u);
    }

protected:
    std::vector<std::pair<int, int>> possible_edges;
};

class UniformGraph : public Graph {
public:
    std::string name;

    UniformGraph(int n, int m) : Graph(n) {
        generate_uniform_edges(m);
        name = "uniform";
    }

    void generate_uniform_edges(int m) {
        std::random_device rd;
        std::mt19937 g(rd());

        if (m > possible_edges.size()) {
            std::cerr << "Error: m > possible_edges.size()\n";
            throw std::invalid_argument("Too many edges requested.");
        }

        std::shuffle(possible_edges.begin(), possible_edges.end(), g);

        for (int i = 0; i < m; ++i) {
            add_edge(possible_edges[i].first, possible_edges[i].second);
        }
    }
};

class BinomialGraph : public Graph {
public:
    std::string name;

    BinomialGraph(int n, double p) : Graph(n) {
        generate_binomial_edges(p);
        name = "binomial";
    }

    void generate_binomial_edges(double p) {
        std::random_device rd;
        std::mt19937 g(rd());
        std::uniform_real_distribution<> dist(0.0, 1.0);

        for (const auto &edge : possible_edges) {
            if (dist(g) < p) {
                add_edge(edge.first, edge.second);
            }
        }
    }
};

bool has_at_least_one_triangle(const Graph &g) {
    for (int v1 = 0; v1 < g.edges.size(); ++v1) {
        for (int v2 : g.edges[v1]) {
            for (int v3 = 0; v3 < g.edges.size(); ++v3) {
                if (std::find(g.edges[v3].begin(), g.edges[v3].end(), v1) != g.edges[v3].end() &&
                    std::find(g.edges[v3].begin(), g.edges[v3].end(), v2) != g.edges[v3].end()) {
                    return true;
                }
            }
        }
    }
    return false;
}

bool is_connected(const Graph &g) {
    std::deque<int> queue{0};
    std::unordered_set<int> visited{0};

    while (!queue.empty()) {
        int v = queue.front();
        queue.pop_front();
        for (int neighbor : g.edges[v]) {
            if (visited.find(neighbor) == visited.end()) {
                queue.push_back(neighbor);
                visited.insert(neighbor);
            }
        }
    }

    return visited.size() == g.n;
}

bool has_half_of_vertices_with_degree_4(const Graph &g) {
    int count = 0;
    for (const auto &edges : g.edges) {
        if (edges.size() >= 4) {
            count++;
        }
    }
    return count >= g.n / 2;
}

bool has_amount_edges_different_than_m(int m, const Graph &g) {
    int edge_count = 0;
    for (const auto &edges : g.edges) {
        edge_count += edges.size();
    }
    return m != edge_count / 2;
}

double binom_n_2(int n) {
    return n * (n - 1) / 2.0;
}

// Assume that other necessary functions and classes (Graph, UniformGraph, BinomialGraph, etc.) are already defined above

void main_loop() {
    std::vector<int> ns_tested = {5};
    for (int i = 10; i < 400; i += 50) {
        ns_tested.push_back(i);
    }

    std::vector<std::pair<std::function<double(int)>, std::string>> possible_ms = {
        {[](int n) { return std::sqrt(n); }, "sqrt(n)"},
        {[](int n) { return n; }, "n"},
        {[](int n) { return 2 * n; }, "2n"},
        {[](int n) { return n * std::log2(n); }, "n*log2(n)"},
        {[](int n) { return (1 / 4.0) * n * n; }, "1/4n^2"}
    };

    std::string print_string;

    for (const auto &[possible_m, m_name] : possible_ms) {
        for (int n : ns_tested) {
            int m = static_cast<int>(possible_m(n));
            if (m > binom_n_2(n)) {
                continue;
            }
            double p = m / binom_n_2(n);

            std::vector<std::function<bool(const Graph &)>> features = {
                has_at_least_one_triangle,
                is_connected,
                has_half_of_vertices_with_degree_4,
                [m](const Graph &g) { return has_amount_edges_different_than_m(m, g); }
            };

            std::cout << "Doing tests for n=" << n << ", m=" << m << std::endl;

            // Loop through graph types using strings
            std::vector<std::string> graph_types = {"uniform", "binomial"};

            const int AMOUNT_TESTS = 100;
            for (const auto& graph_type : graph_types) {
                // Store results for each feature
                std::vector<double> feature_results(features.size(), 0.0);

                // Run the tests and generate a new graph for each iteration
                for (int i = 0; i < AMOUNT_TESTS; ++i) {
                    std::shared_ptr<Graph> graph;

                    // Create a new graph for each test iteration
                    if (graph_type == "uniform") {
                        graph = std::make_shared<UniformGraph>(n, m);
                    } else if (graph_type == "binomial") {
                        graph = std::make_shared<BinomialGraph>(n, p);
                    }

                    // Evaluate all features for the generated graph
                    for (size_t j = 0; j < features.size(); ++j) {
                        feature_results[j] += features[j](*graph);
                    }
                }

                // Calculate averages
                for (auto &result : feature_results) {
                    result /= AMOUNT_TESTS;
                }

                // Append results to print_string
                print_string += std::to_string(n) + " " + m_name + " " + graph_type + " ";
                for (double result : feature_results) {
                    print_string += std::to_string(result) + " ";
                }
                print_string += "\n";
            }
        }
    }

    // Write results to file
    std::ofstream file("data.txt");
    if (file.is_open()) {
        file << print_string;
        file.close();
    } else {
        std::cerr << "Unable to open file data.txt\n";
    }
}

int main() {
    main_loop();
    return 0;
}

File: glisz/misja1/mission1.py
import cProfile
from collections import defaultdict
import json
import math
import random
from collections import deque

class Graph:
    edges: list[list[int]]
    n: int
    
    def __init__(self, n: int) -> None:
        self.n = n
        self.possible_edges = [(u,v) for u in range(self.n) for v in range(u+1, self.n)]
        self.edges = [[] for _ in range(self.n)]
        
    def add_edge(self, u, v):
        self.edges[u].append(v)
        self.edges[v].append(u)
        
class UniformGraph(Graph):
    def __init__(self, n: int, m: int) -> None:
        super().__init__(n)
        self.generate_uniform_edges(m)
        self.name = "uniform"
        
    def generate_uniform_edges(self, m: int) -> None:
        try:
            for u, v in random.sample(self.possible_edges, k=m):
                self.add_edge(u, v)
        except ValueError:
            print(f"Error: {self.possible_edges=}, {m=}")
            raise

        

class BinomialGraph(Graph):
    def __init__(self, n: int, p: float) -> None:
        super().__init__(n)
        self.generate_binomial_edges(p)
        self.name = "binomial"

    def generate_binomial_edges(self, p: float):
        for u, v in self.possible_edges:
            if random.random() < p:
                self.add_edge(u, v)


def has_at_least_one_triangle(g: Graph) -> bool:
    for v1 in range(len(g.edges)):
        for v2 in g.edges[v1]:
            for v3 in range(len(g.edges)):
                if v1 in g.edges[v3] and v2 in g.edges[v3]:
                    return True
    return False

def is_connected(g: Graph) -> bool:
    visited = set([0])
    queue = deque([0])
    while queue:
        v = queue.popleft()
        for neigh in g.edges[v]:
            if neigh not in visited:
                queue.append(neigh)
                visited.add(neigh)
    return len(visited) == g.n

def has_half_of_vertices_with_degree_4(g: Graph) -> bool:
    amount_of_vertices_with_degree_4 = 0
    for edges in g.edges:
        if len(edges) >= 4:
            amount_of_vertices_with_degree_4 += 1
    return amount_of_vertices_with_degree_4 >= g.n / 2

def has_amount_edges_different_than_m(m: int, g: Graph) -> bool:
    return m != (sum(len(edges) for edges in g.edges) / 2)

def binom_n_2(n: int) -> float:
    return n * (n-1) / 2


def main():
    ns_tested = [5] + list(range(10, 400, 50))
    possible_ms = [
        (lambda n: math.sqrt(n), "sqrt(n)"),
        (lambda n: n, "n"),
        (lambda n: 2*n, "2n"),
        (lambda n: n * math.log2(n), "n*log2(n)"),
        (lambda n: (1/4) * n**2, "1/4n**2")
    ]
    data = dict()
    print_string = ""
    
    for possible_m, m_name in possible_ms:
        for n in ns_tested:
            m = int(possible_m(n))
            if m > binom_n_2(n):
                continue
            p = m / binom_n_2(n)
            
            features = [
                has_at_least_one_triangle,
                is_connected,
                has_half_of_vertices_with_degree_4,
                lambda g: has_amount_edges_different_than_m(m, g)
            ]
            
            print(f"Doing tests for {n=}, {m=}")
            for g in [lambda: UniformGraph(n, m), lambda: BinomialGraph(n, p)]:
                
                AMOUNT_TESTS = 100
                for _ in range(AMOUNT_TESTS):
                    graph = g()
                    graph_features = [f(graph) for f in features]
                    
                
                    if n not in data:
                        data[n] = dict()
                    if m_name not in data[n]:
                        data[n][m_name] = dict()
                        
                    if g not in data[n][m_name]:
                        data[n][m_name][graph.name] = graph_features
                    else:
                        for i in range(len(graph_features)):
                            data[n][m_name][graph.name][i] += graph_features[i]
                
                for i in range(len(data[n][m_name][graph.name])):
                    data[n][m_name][graph.name][i] /= AMOUNT_TESTS
                
                print_string += f"{n} {m_name} {g} {' '.join([str(x) for x in data[n][m_name][graph.name]])}\n"
    
    with open("data.txt", 'w+') as file:
        file.write(print_string)
                
if __name__ == "__main__":
    main()